akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala

```diff
@@ -534,6 +535,93 @@ final class MergeSorted[T: Ordering] extends GraphStage[FanInShape2[T, T, T]] {
   }
 }
 
+object MergeSortedN {
+  /**
+   * Create a new `MergeSortedN`.
+   */
+  def apply[A: Ordering](n: Int) = new MergeSortedN[A](n)
+}
+
+/**
+ * Merge multiple pre-sorted streams such that the resulting stream is sorted.
+ *
+ * '''Emits when''' all inputs have an element available
+ *
+ * '''Backpressures when''' downstream backpressures
+ *
+ * '''Completes when''' all upstreams complete
+ *
+ * '''Cancels when''' downstream cancels
+ */
+final class MergeSortedN[T: Ordering](inputPorts: Int) extends GraphStage[UniformFanInShape[T, T]] {
+  val inlets: immutable.IndexedSeq[Inlet[T]] = Vector.tabulate(inputPorts)(i ⇒ Inlet[T]("MergeSortedN.in" + i))
+  val out: Outlet[T] = Outlet[T]("MergeSortedN.out")
+  override val shape: UniformFanInShape[T, T] = UniformFanInShape(out, inlets: _*)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) with OutHandler {
+    private val bufferedElements = new mutable.TreeSet[(T, Int)]()
+    private val inletsBeingPulled = new mutable.HashSet[Inlet[T]]
+    private var runningInlets = inputPorts
+    private def canPush = inletsBeingPulled.isEmpty && bufferedElements.nonEmpty
+    private def upstreamsClosed = runningInlets == 0
+
+    override def preStart(): Unit = {
+      for (i ← inlets) {
+        tryPull(i)
+        inletsBeingPulled.add(i)
+      }
+    }
+
+    private def maybePush(): Unit = {
+      if (canPush) {
+        val next = bufferedElements.firstKey
+        bufferedElements.remove(next)
+        push(out, next._1)
+        val inlet = inlets(next._2)
+        if (!isClosed(inlet)) {
+          pull(inlet)
+          inletsBeingPulled.add(inlet)
+        }
+      }
+
+      if (bufferedElements.isEmpty && upstreamsClosed) {
+        completeStage()
+      }
+    }
+
+    for (i ← inlets.indices) {
+      val inlet = inlets(i)
+      val index = i
+      setHandler(inlet, new InHandler {
+        override def onPush(): Unit = {
+          val element = grab(inlet)
+          bufferedElements.add((element, index))
```

Comment:
Do we need any bounds on this buffer?
Do I understand it correctly that we don't pull from that inlet until we have received one element from each inlet? Thereby the buffer is bounded.

Reply:
the buffer is indeed bounded. It should contain a maximum of one element per inlet